### PR TITLE
MODINVOSTO-34 Implement API for invoice attachments (links/documents)

### PR DESCRIPTION
**Pull request by grabsefx**
## Purpose
https://issues.folio.org/browse/MODINVOSTO-34

## Approach
* RAML updated
* The API is implemented (POST, GET by ID, GET by Query, Delete). Implemented PUT endpoint to always respond with code 501 - NOT_IMPLEMENTED
* Unit tests updated
* endpoint permissions added

## TODO 
* merge https://github.com/folio-org/acq-models/pull/168 and update acq-models submodule

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [x] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by dcrossleyau** [comment]
In this case, not sure why the "lint-raml" message is not more detailed.
Locally using "z-schema" directly (one of the tools underlying lint-raml) shows `... not a valid RegExp pattern` for this beauty:

ramls/acq-models/mod-invoice-storage/schemas/document_metadata.json:
```
"pattern": "(?i)\\b((?:[a-z][\\w-]+:(?:/{1,3}|[a-z0-9%])|www\\d{0,3}[.]|[a-z0-9.\\-]+[.][a-z]{2,4}/)(?:[^\\s()<>]+|\\(([^\\s()<>]+|(\\([^\\s()<>]+\\)))*\\))+(?:\\(([^\\s()<>]+|(\\([^\\s()<>]+\\)))*\\)|[^\\s`!()\\[\\]{};:'\".,<>?]))"

**Comment by dcrossleyau** [comment]
Ah, found a new favourite tool ... https://regexr.com

It refers to the front portion ...
```
(?i) ERROR: The "mode modifier" feature is not supported in this flavor of RegEx
```

So perhaps "z-schema" does not either.

That is a scary topic, url regex.

If this one is required, then i can only suggest turning off the "runLintRamlCop" in Jenkinsfile.

File: src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java [hunk 1]
```diff
@@ -158,6 +174,140 @@ public void putInvoiceStorageInvoicesById(String id, String lang, Invoice entity
...
+          String tenantId = TenantTool.calculateTenantId(okapiHeaders.get(RestVerticle.OKAPI_HEADER_TENANT));
+          String fullTableName = PostgresClient.convertToPsqlStandard(tenantId) + "." + DOCUMENT_TABLE;
+
+          entity.getDocumentMetadata().setMetadata(entity.getMetadata());
+
+          boolean base64Exists = entity.getContents() != null && StringUtils.isNotEmpty(entity.getContents().getData());
+          String sql = "INSERT INTO " + fullTableName + " (id, " + (base64Exists ? "document_data," : "") + " jsonb) VALUES (?," + (base64Exists ? "?," : "") + " ?::JSON) RETURNING id";
+          JsonArray jsonArray = prepareDocumentQueryParams(entity);
+
+          sqlConnection.result().queryWithParams(sql, jsonArray, query -> {
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: Please consider to use `pgClient.execute(sqlConnection, sql, jsonArray, <replyHandler>)`

File: pom.xml [hunk 2]
```diff
@@ -13,13 +13,14 @@
     </licenses>
 
   <properties>
-    <raml-module-builder.version>26.1.0</raml-module-builder.version>
+    <raml-module-builder.version>26.2.0</raml-module-builder.version>
```
Comment (javalkar-varun) [inline pom.xml]: A newer version is available  `v26.2.4`

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -33,6 +33,37 @@
...
+          "pathPattern": "/invoice-storage/invoices{id}/documents",
+          "permissionsRequired": ["invoice-storage.invoices.documents.item.post"]
+        },
+        {
+          "methods": ["GET"],
+          "pathPattern": "/invoice-storage/invoices/{id}/documents/{documentId}",
+          "permissionsRequired": ["invoice-storage.invoices.documents.item.get"]
+        },
+        {
+          "methods": ["PUT"],
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply descriptors/ModuleDescriptor-template.json]: removed PUT endpoint at all

File: ramls/invoice.raml [hunk 4]
```diff
@@ -55,6 +57,30 @@ resourceTypes:
           schema: invoice
           exampleItem: !include acq-models/mod-invoice-storage/examples/invoice.sample
 
+      /documents:
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply ramls/invoice.raml]: Updated description

File: src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java [hunk 5]
```diff
@@ -158,6 +174,140 @@ public void putInvoiceStorageInvoicesById(String id, String lang, Invoice entity
...
 
+  @Validate
+  @Override
+  public void getInvoiceStorageInvoicesDocumentsById(String id, String lang, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    log.info("Get document list by invoice id");
+    vertxContext.runOnContext((Void v) -> {
+      try {
+        Criterion criterion = getCriterionByFieldNameAndValue(INVOICE_ID_FIELD_NAME, id);
+        pgClient.get(DOCUMENT_TABLE, Document.class, criterion, false, reply -> {
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: Implementation was updated to use `PgUtil.get(<10 args>)`

[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: Implementation was updated to use `PgUtil.get(<10 args>)`

File: src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java (hunk 1 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: Implementation was updated to use `pgClient.execute`

File: src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java [hunk 6]
```diff
@@ -158,6 +173,106 @@ public void putInvoiceStorageInvoicesById(String id, String lang, Invoice entity
     });
   }
 
+  @Validate
+  @Override
+  public void getInvoiceStorageInvoicesDocumentsById(String id, String lang, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+
+    String cqlQuery = INVOICE_ID_FIELD_NAME + "=" + id;
+    PgUtil.get(DOCUMENT_TABLE, DocumentMetadata.class, DocumentCollection.class, cqlQuery, 0, 999, okapiHeaders, vertxContext,
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: As an alternative approach `pageable` trait can be used so client can define how many records are required

Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: implemented this endpoint to support sort and filtering

File: src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java [hunk 7]
```diff
@@ -158,6 +173,106 @@ public void putInvoiceStorageInvoicesById(String id, String lang, Invoice entity
...
+    );
+  }
+
+  @Validate
+  @Override
+  public void getInvoiceStorageInvoicesDocumentsByIdAndDocumentId(String id, String documentId, String lang,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    vertxContext.runOnContext((Void v) -> {
+      try {
+        String query = "SELECT jsonb, document_data FROM " + DOCUMENT_TABLE + " WHERE id ='" + documentId + "'";
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: fixed

File: ramls/invoice.raml [hunk 8]
```diff
@@ -55,6 +58,32 @@ resourceTypes:
...
+            exampleCollection: !include acq-models/mod-invoice-storage/examples/document_collection.sample
+            exampleItem: !include acq-models/mod-invoice-storage/examples/document.sample
+        get:
+          description: Get list of documents
+          is: [
+            pageable,
+            searchable: {description: "with valid searchable fields: for example metadata.createdDate", example: "metadata.createdDate > '2018-07-19T00:00:00.000+0000'"}
+          ]
+        /{documentId}:
+          description: Get and Delete a specific invoice. Update operation is not allowed
```
Comment (craigmcnally) [inline ramls/invoice.raml]: probably want to fix this description... we're not getting/deleting invoices, but invoice document attachments/links

Comment (grabsefx) [reply ramls/invoice.raml]: fixed

File: src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java [hunk 9]
```diff
@@ -158,6 +159,111 @@ public void putInvoiceStorageInvoicesById(String id, String lang, Invoice entity
...
+  @Validate
+  @Override
+  public void getInvoiceStorageInvoicesDocumentsById(String id, int offset, int limit, String query, String lang,
+      Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    String getByIdQuery = INVOICE_ID_FIELD_NAME + "==" + id;
+    String resultQuery = StringUtils.isNotEmpty(query) ? combineCqlExpressions("and", getByIdQuery, query) : getByIdQuery;
+
+    EntitiesMetadataHolder<Document, DocumentCollection> entitiesMetadataHolder = new EntitiesMetadataHolder<>(Document.class, DocumentCollection.class, GetInvoiceStorageInvoicesDocumentsByIdResponse.class);
+    QueryHolder cql = new QueryHolder(DOCUMENT_TABLE, resultQuery, offset, limit, lang);
+    getEntitiesCollection(entitiesMetadataHolder, cql, asyncResultHandler, vertxContext, okapiHeaders);
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/impl/InvoiceStorageImpl.java]: You can use `PgUtils.get(...)` instead of  this.

**Pull request merged**